osu.Server.Queues.ScoreStatisticsProcessor/Commands/Performance/Scores/UpdateAllScoresCommand.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Data;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dapper;
+using McMaster.Extensions.CommandLineUtils;
+using osu.Server.QueueProcessor;
+using osu.Server.Queues.ScoreStatisticsProcessor.Helpers;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Commands.Performance.Scores
+{
+    [Command(Name = "all", Description = "Computes pp of all scores from all users.")]
+    public class UpdateAllScoresCommand : PerformanceCommand
+    {
+        private const int max_scores_per_query = 5000;
+
+        [Option(Description = "Score ID to start processing from.")]
+        public ulong From { get; set; }
+
+        /// <summary>
+        /// Whether to adjust processing rate based on slave latency. Defaults to <c>false</c>.
+        /// </summary>
+        [Option(CommandOptionType.SingleOrNoValue, Template = "--check-slave-latency")]
+        public bool CheckSlaveLatency { get; set; }
+
+        protected override async Task<int> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            using var db = DatabaseAccess.GetConnection();
+
+            ulong currentScoreId = From;
+            ulong? lastScoreId = await db.QuerySingleAsync<ulong>("SELECT MAX(id) FROM scores");
+
+            ulong processedCount = 0;
+            ulong changedPp = 0;
+            double rate = 0;
+            Stopwatch sw = new Stopwatch();
+
+            var scoresQuery = db.Query<SoloScore>("SELECT * FROM scores WHERE `id` > @ScoreId AND `id` <= @LastScoreId ORDER BY `id`", new
+            {
+                ScoreId = currentScoreId,
+                LastScoreId = lastScoreId,
+            }, buffered: false);
+
+            using var scoresEnum = scoresQuery.GetEnumerator();
+
+            Console.WriteLine($"Processing all scores up to {lastScoreId}, starting from {currentScoreId}");
+
+            Task<List<SoloScore>> nextScores = getNextScores();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (CheckSlaveLatency)
+                    await SlaveLatencyChecker.CheckSlaveLatency(db, cancellationToken);
+
+                sw.Restart();
+
+                var scores = await nextScores;
+                nextScores = getNextScores();
+
+                if (scores.Count == 0)
+                    break;
+
+                await Task.WhenAll(Partitioner.Create(scores).GetPartitions(Threads).Select(async partition =>

```

Comment:
Any reason for not using `ProcessPartitioned`?

Reply:
I couldn't make the transaction work easily when using that, so had to inline it.